File: leetcode/pascal_triangle/PascalTriangle.cpp
class Solution {
public:
    vector<vector<int> > generate(int numRows) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        vector<vector<int> > result; 
        if (numRows == 0) {
            return result; 
        }
        
        vector<int> row; 
        row.push_back(1);
        result.push_back(row); 
        
        for (int i = 1; i < numRows; i++) {
            row.clear();
            row.push_back(1); 
            vector<int> currRow = result[i - 1]; 
            for (int j = 1; j < currRow.size(); j++) {
                int sum = currRow[j - 1] + currRow[j]; 
                row.push_back(sum); 
            }
            row.push_back(1); 
            result.push_back(row); 
        }
        
        return result; 
    }
};

File: leetcode4/PascalTriangle.java
public class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer> > result = new ArrayList<List<Integer> >(); 
        
        if (numRows <= 0) {
            return result; 
        }
        
        for (int i = 0; i < numRows; i++) {
            if (i == 0) {
                result.add(new ArrayList<Integer>(Arrays.asList(1))); 
            } else {
                List<Integer> prev = result.get(i - 1);
                List<Integer> curr = new ArrayList<Integer>(); 
                curr.add(1);
                
                for (int j = 0; j < prev.size() - 1; j++) {
                    curr.add(prev.get(j) + prev.get(j + 1)); 
                }
                
                curr.add(1); 
                result.add(curr); 
            }
        }
        
        return result; 
    }
}
